kubectl-volsync/cmd/migration_create.go

```diff
@@ -36,70 +93,309 @@ var migrationCreateCmd = &cobra.Command{
+
+	serviceType, err := cmd.Flags().GetString("servicetype")
+	if err != nil {
+		return fmt.Errorf("please provide service type, err = %w", err)
+	}
+
+	if v1.ServiceType(serviceType) != v1.ServiceTypeClusterIP &&
+		v1.ServiceType(serviceType) != v1.ServiceTypeLoadBalancer {
+		return fmt.Errorf("unsupported service type: %v", v1.ServiceType(serviceType))
+	}
+	mc.ServiceType = (*v1.ServiceType)(&serviceType)
+	mc.RDName = mc.Namespace + "-" + mc.PVCName + "-migration-dest"
+
+	return nil
+}
+
+//nolint:funlen
+func (mc *migrationCreate) newMigrationRelationshipDestination(ctx context.Context) error {
+	mrd := mc.mr.data.Destination
+
+	// Assign the values from migrationCreate built after parsing cmd args
+	mrd.RDName = mc.RDName
+	mrd.PVCName = mc.PVCName
+	mrd.Namespace = mc.Namespace
+	mrd.Cluster = mc.Cluster
+
+	if mc.PVC = mc.getDestinationPVC(ctx); mc.PVC == nil {
+		if mc.Capacity == nil {
+			return fmt.Errorf("capacity arg must be provided")
+		}
+		mrd.Destination.Capacity = mc.Capacity
+
+		// Default "AccessMode = ReadWriteOnce" if not provided
+		mrd.Destination.AccessModes = mc.AccessModes
+
+		if mc.StorageClassName == nil {
+			klog.Infof("Storage class is not specified as part of cmd arg, binding to default storage class")
+			mrd.Destination.StorageClassName = nil
+		} else {
+			mrd.Destination.StorageClassName = mc.StorageClassName
+		}
+
+		mrd.Destination.DestinationPVC = &mc.PVCName
+	} else {
+		mrd.Destination.DestinationPVC = &mc.PVC.Name
+		mrd.Cluster = mc.PVC.ClusterName
 	}
+
+	mrd.Destination.CopyMethod = mc.CopyMethod
+	mrd.Destination.ServiceType = mc.ServiceType
+
 	return nil
 }
 
-func doMigrationCreate(cmd *cobra.Command, args []string) error {
-	// Create the empty relationship
-	configDir, err := cmd.Flags().GetString("config-dir")
+func (mc *migrationCreate) Run(ctx context.Context) error {
+	mrd := &migrationRelationshipDestination{}
+	mc.mr.data.Destination = mrd
+
+	k8sClient, err := newClient(mrd.Cluster)
 	if err != nil {
 		return err
 	}
-	rName, err := cmd.Flags().GetString("relationship")
+	mc.clientObject = k8sClient
+
+	// build struct migrationRelationshipDestination from struct migrationCreate
+	err = mc.newMigrationRelationshipDestination(ctx)
 	if err != nil {
 		return err
 	}
-	relation, err := createRelationship(configDir, rName, MigrationRelationship)
+	// create namespace if does not exist
+	_, err = mc.createNamespace(ctx)
 	if err != nil {
 		return err
 	}
-
-	// Insert information into the relationship & save it
-	cap, err := cmd.Flags().GetString("capacity")
+	// create destination PVC if does not exist
+	if mc.PVC == nil {
+		mc.PVC, err = mc.createDestinationPVC(ctx)
+		if err != nil {
+			return err
+		}
+	} else {
+		klog.Infof("Destination PVC: \"%s\" is found in Namespace: \"%s\" and is used to create replication destination",
+			mc.PVC.Name, mc.PVC.Namespace)
+	}
+	// create migration destination
+	err = mc.createDestination(ctx)
 	if err != nil {
 		return err
 	}
```

Comment:
I suggest a slightly different approach that's similar to how the operator itself works...
If you change the ordering:
- Namespace
- PVC
- ReplicationDestination

...you'll have the info available as you need it.

Then for the functions that do the work, think of them as "ensuring the state is a certain way". So consider:
```go
// Creates the Namespace if it doesn't exist
func (mc *migrationCreate) ensureNamespace (ctx context.Context) (*corev1.Namespace, error) {
...
}

// Creates the PVC if it doesn't exist
func (mc *migrationCreate) ensureDestPVC (ctx context.Context) (*corev1.PersistentVolumeClaim, error) {
...
}

// Creates the RD if it doesn't exist
func (mc *migrationCreate) ensureReplicationDestination (ctx context.Context) (*volsyncv1alpha1.ReplicationDestination, error) {
...
}
```

Then Run() becomes just calling them in sequence. And you know when writing, for example, ensureReplicationDestination that the PVC is there, and you have a pointer to it in mc.PVC.

Reply:
sure